### DISCUSSION TITLE
Use on React Native

### DISCUSSION THREAD

**Discussion by poltak** [openpgpjs/openpgpjs#1211] [Q&A]
Hi there.

Has anyone had success getting this lib to work on React Native?

Adding it straight into the RN project I'm working on immediately crashes the app at load, saying it can't resolve the `crypto` module.

Looking deeper into this package I found that both the use of `crypto` and `streams` APIs seem to not be available on RN.

For `crypto`, there is a RN port here: https://github.com/tradle/react-native-crypto
For `streams`, `readable-stream` seems RN-compatible: https://github.com/nodejs/readable-stream

I ended up creating a fork of this repo, replacing the importing of the above libs with the RN-compatible ones. 

This got me past some of these errors, however I realized the `react-native-crypto` leads to an error when running the NPM `build` script, which I presume is something to do with the current grunt setup:

```
$ grunt build --compat copy:openpgp_compat && grunt build --lightweight copy:openpgp_lightweight clean:js && grunt build
Running "build" task

Running "browserify:openpgp" (browserify) task
>> Error: Parsing file ~/Documents/github/openpgpjs/node_modules/react-native-crypto/index.js: 'import' and 'export' may appear only with 'sourceType: module' (3:0)
Warning: Error running grunt-browserify. Use --force to continue.
```

I also saw some of the code to switch between using certain deps depending on whether or not it is being run in node or the browser. Perhaps adding proper support for React Native would also need an update to this code to factor in a environment.

I will add to this with further findings.

What else does anyone know that is stopping this from just working with React Native?

**Comment by poltak** [comment]
I ended up getting past the build error by adding `react-native-crypto` to the `browserify.openpgp.options.external` array so that package doesn't get bundled in the build. Also using NPM rather than yarn fxied a follow-up build issue (#926).

I'm now however stuck with the importing of the `buffer` package crashing the app at runtime. I ended up updating the `getNodeBuffer` function in `util.js` to return `undefined`, however the issue persists with that build. May be it's also be required by another dep. I've also tried adding in https://github.com/feross/buffer, without any luck

**Comment by twiss** [comment]
Hey @poltak :wave: I don't know much about React Native, but if it behaves more like a browser than like Node.js, you might have better luck trying to convince OpenPGP.js that it's running in a browser rather than that it's running in Node.js. Probably you could do this by replacing `window` with `global` everywhere in OpenPGP.js (PR welcome), or creating a fake `global.window = global` variable before loading OpenPGP.js.

**Comment by poltak** [comment]
From what I understand, it behaves different to both node and browser. You're missing lots of APIs that are available in both/either node and browsers. There seem to be lots of ports of certain APIs available for React Native, like with the streams, crypto, and buffer libs that I've linked. Unfortunately I haven't got those all working properly with this implementation of OpenPGP.

There are some other implementations of OpenPGP available for React Native, though they don't seem to be up to this project's standard of being tried-and-true. So far I've found these ones:
- https://github.com/jerson/react-native-fast-openpgp
- https://github.com/orhan/react-native-openpgp

> Probably you could do this by replacing window with global everywhere in OpenPGP.js (PR welcome), or creating a fake global.window = global variable before loading OpenPGP.js.

Unfortunately this didn't have any effect, but thanks for the suggestion.

**Comment by gabidi** [comment]
Hey @poltak , Did you ever manage to resolve this issue ?
Like you said other libs are available but non come close to the level of this one.
Would appreciate your thoughts, thanks.

**Comment by ken0x0a** [comment]
@poltak @gabidi  First of all, as I know, you will end up with 1 binary (index.bundle or something else) if you are using React Native, and the license of this library is LGPL, so it will be required to open source all of your js code with LGPL.

Could you let me know if you know any workaround for this.

I also wanted to use this functionality in RN, but I couldn't as the license.

**Comment by ken0x0a** [comment]
@openpgpjs
Thank you for this great library!

I want to use this library at React Native, but as above reason, I cannot use right now.

Is there any chance for you to change LICENSE?
Is it impossible?

**Comment by gabidi** [comment]
@ken0x0a Open sourcing the application is part of our product roadmap and is not an issue for me. Thank you for taking the time to raise that flag, I do appreciate it.

**Comment by chesnokovilya** [comment]
The library LGPL license is not preventing it from being used inside proprietary software:
"A program that contains no derivative of any portion of the Library, but is designed to work with the Library by being compiled or linked with it, is called a "work that uses the Library". Such a work, in isolation, is not a derivative work of the Library, and therefore falls outside the scope of this License."

**Comment by ken0x0a** [comment]
@chesnokovilya Thank you for your answer!

Can I ask further more about it?

What you told me means:
- a work will be LGPL, if that is similar with this library;
- a work will not be LGPL, if that using this library for only some functionality of that work, and not the main functionality;

Is it correct?

**Comment by chesnokovilya** [comment]
I think it is not correct. 
Here is good explanation from wikipedia [https://en.wikipedia.org/wiki/GNU_Lesser_General_Public_License]
Essentially, if it is a "work that uses the library", then it must be possible for the software to be linked with a newer version of the LGPL-covered program. The most commonly used method for doing so is to use "a suitable shared library mechanism for linking". Alternatively, a statically linked library is allowed if either source code or linkable object files are provided.

**Comment by ken0x0a** [comment]
@chesnokovilya Thank you for the reply!

As I understand, I can use this library for my work.
But also, I understand that the definition of ["derivative work"](https://en.wikipedia.org/wiki/Derivative_work) differs for each person, and it's not clear whether it's safe to use for non-public work or not.

As I found another libraries with clear license, so I should use them.

Thank you so much for your help and time 🙏

**Comment by lauhon** [comment]
I got this library to work together with react-native by doing some steps. Right now my solution will only work by modifying a library which is needed. I opened [this PR](https://github.com/tradle/react-native-crypto/pull/69) to address that.

Basically, my solution is to use [https://www.npmjs.com/package/rn-nodeify](rn-nodeify) to make sure react-native has (almost) every API as normal Node.js. To get it working you can follow these steps:

1. `npm i rn-nodeify`

2. Add a new npm script like the following  in your `package.json` 
```
  "scripts": {
    "nodeify": "rn-nodeify --install buffer,stream,assert,events,crypto,vm,process,zlib,os --hack"
  }

```
(It might be that you dont actually need all of the listed shims, this is simply what I have in my project)


3. OPTIONAL: I added this new script to "postinstall", so it gets executed every time I run `npm install`automatically
```
  "scripts": {
    "postinstall": "npm run nodeify",
    "nodeify": "..."
  }

```
At this point you already have every needed Node Functionality available. But there still is one catch, this is where my PR mentioned above comes in to play.

4. `rn-nodeify` uses [react-native-crypto](https://github.com/tradle/react-native-crypto) to provide a `crypto` API. In this package a function called `getCurves()` is missing. This function is added by my PR. For a development environment it is enough (at least for me) to add the method locally inside the node_modules folder.

Go to `node_modules/react-native-crypto/index.js`and add following line at the end
`exports.getCurves = () => [];`

5. One last thing is to make sure you actually use the node version of this package in react-native. I am importing openpgp like this:
`const openpgp = require("openpgp/dist/node/openpgp.min.js");`
Of course you can also import this, but you might get a typescript warning, depending on your tsconfig

**Comment by twiss** [comment]
Hi :wave: Thanks for looking into it! It seems this uses the browserify-crypto implementation of the Node.js crypto API in pure JS. For performance and security, I would recommend using the native web crypto API instead, if possible. [react-native-webview-crypto](https://www.npmjs.com/package/react-native-webview-crypto) seems to be an attempt at that (though the interface seems to add some overhead, so I'm not sure if it's actually more performant, but it should at least be constant-time). You could try using that, together with [react-native-get-random-values](https://www.npmjs.com/package/react-native-get-random-values), and then use the browser build of OpenPGP.js (instead of the Node.js build). That seems like a more straightforward way to go, but I haven't tested it, so I'm not sure if it works.

The other advantage of that would be that you'd get the asmcrypto.js fallbacks for algorithms that aren't natively supported in Web Crypto, rather than the browserify-crypto implementations. The former are probably more performant, but again, I haven't benchmarked the latter so YMMV.

**Reply by lauhon** [reply]
Hi, thanks for the response.

It seems my comment wasnt the End of the Story. The Library could do _some_ things in the react native context after my workaround, but I still got errors when I wanted to publicEncrypt for example. So I'm considering what you are writing.

My other thought is to just skip openpgp on the client and find another encryption mechanism (maybe symmetric) to send the information to my node.js backend which then can just use openpgpjs as intended.

**Reply by twiss** [reply]
Sure. If you don't need OpenPGP compatibility, you can just use symmetric crypto on both ends. Web Crypto (using [react-native-webview-crypto](https://www.npmjs.com/package/react-native-webview-crypto)) still seems like a reasonable option for that, as [Node.js has an experimental implementation of that as well](https://nodejs.org/api/webcrypto.html). Alternatively, you could use something like [TweetNaCl.js](https://tweetnacl.js.org/) (together with [react-native-get-random-values](https://www.npmjs.com/package/react-native-get-random-values)).
